internal/docker/podman.go

```diff
@@ -0,0 +1,236 @@
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/anchore/stereoscope/internal/log"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh"
+)
+
+var (
+	// Podman supports the following connections
+	// rootless "unix://run/user/$UID/podman/podman.sock" (Default)
+	// rootfull "unix://run/podman/podman.sock (Default)
+	// remote rootless ssh://engineering.lab.company.com/run/user/$UID/podman/podman.sock
+	// remote rootfull ssh://root@10.10.1.136:22/run/podman/podman.sock
+
+	localRootlessPath     = getLocalRootlesPath()
+	defaultRemoteRootless = fmt.Sprintf("ssh://core@localhost:63753%s?secure=false", localRootlessPath)
+	defaultLocalRootless  = fmt.Sprintf("unix://%s", localRootlessPath)
+)
+
+func getLocalRootlesPath() string {
+	return fmt.Sprintf("unix:///run/user/%d/podman/podman.sock", os.Getuid())
+}
+
+func podmanOverSSH() (*client.Client, error) {
+	log.Debug("using docker client to connect to podman over ssh")
+	var clientOpts = []client.Opt{
+		client.WithAPIVersionNegotiation(),
+		// TODO: is this needed?
+		client.FromEnv,
+	}
+
+	host := defaultRemoteRootless
+	makeClient := sshClient
+
+	if v, found := os.LookupEnv("CONTAINER_HOST"); found && v != "" {
+		log.Debugf("using $CONTAINER_HOST: %s", v)
+		host = v
+	}
+
+	if v, found := os.LookupEnv("PODMAN_HOST"); found && v != "" {
+		log.Debugf("using $PODMAN_HOST: %s", v)
+		host = v
+	}
+
+	log.Debugf("host: %q", host)
+
+	hostURL, err := url.Parse(host)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing host %s with: %w", host, err)
+	}
+
+	httpClient, err := makeClient(hostURL)
+	if err != nil {
+		return nil, fmt.Errorf("making http client: %w", err)
+	}
+
+	clientOpts = append(clientOpts, func(c *client.Client) error {
+		return client.WithHTTPClient(httpClient)(c)
+	})
+	clientOpts = append(clientOpts, client.WithHost("http://d"))
+
+	dockerClient, err := client.NewClientWithOpts(clientOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed create remote client for podman: %w", err)
+	}
+
+	return dockerClient, err
+}
+
+func podmanViaUnixSocket() (*client.Client, error) {
+	log.Debug("using docker client to connect to podman over local unix socket")
+	var clientOpts = []client.Opt{
+		client.WithAPIVersionNegotiation(),
+		// TODO: is this needed?
+		client.FromEnv,
+	}
+
+	// TODO: note this is shared code
+	//if v, found := os.LookupEnv("PODMAN_HOST"); found && v != "" {
+	//	log.Debugf("using $PODMAN_HOST: %s", v)
+	//	host = v
+	//}
+
+	// TODO: if DOCKER_HOST is configued will this take precidence?
+	clientOpts = append(clientOpts, client.WithHost(getLocalRootlesPath()))
+
+	dockerClient, err := client.NewClientWithOpts(clientOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed create local client for podman: %w", err)
+	}
+
+	return dockerClient, err
+}
+
+func GetClientForPodman() (*client.Client, error) {
```

Comment:
this detection could probably be based on the existence of a podman socket at a predetermined path instead of by only OS since podman over SSH is a valid setup on linux. Maybe a hybrid approach where OS is used for windows/darwin, but the linux path has an extra conditional?